### PR TITLE
Fauna Loader

### PR DESCRIPTION
**Pull request by Shadid12**
### How to use it 

```
from langchain.document_loaders.fauna import FaunaLoader

loader = FaunaLoader("Item.all()", "<Field to Retrieve>", "<Fauna Secret>")
docs = loader.load()
print(docs)
```

### CODE REVIEW

File: langchain/document_loaders/fauna.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        self.secrect = secrect
+        self.metadata_fields = metadata_fields
+
+    def load(self) -> List[Document]:
+        try:
+            from fauna import fql, Page
+            from fauna.client import Client
+            from fauna.encoding import QuerySuccess
+        except ImportError:
+            raise ValueError(
```
Comment (leo-gan) [inline langchain/document_loaders/fauna.py:27]: should be `raise ImportError(...`

**Comment by leo-gan** [comment]
Is it possible to add some integration tests and an example into `docs/modules/indexes/document_loaders/examples/`? Thanks!

**Comment by hwchase17** [comment]
+1 to @leo-gan 's comment, thanks for adding this! would love to get this in thursday, all thats missing is an example notebook

File: langchain/document_loaders/fauna.py (hunk 1 above)
Comment (Shadid12) [reply langchain/document_loaders/fauna.py]: updated

File: langchain/document_loaders/fauna.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                page_content = ''
+                for key, value in document_dict.items():
+                    if key == self.page_content_field:
+                        page_content = value
+                document: Document = Document(
+                    page_content=page_content,
+                    metadata={"id": result.id, "ts": result.ts},
+                )
+                documents.append(document)
+        if page.after is not None:
```
Comment (dev2049) [inline langchain/document_loaders/fauna.py:49]: is this saying there's pagination? like if you performed the query again you'd get more/new documents?

File: langchain/document_loaders/fauna.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            )
+        documents = []
+        # Create Fauna Client
+        client = Client(secret=self.secrect) 
+        # Run FQL Query
+        response: QuerySuccess = client.query(fql(self.query))
+        page: Page = response.data
+        for result in page:
+            if result is not None:
+                document_dict = dict(result.items())
```
Comment (dev2049) [inline langchain/document_loaders/fauna.py:39]: why is this needed? what's the type of result?

File: langchain/document_loaders/fauna.py (hunk 2 above)
Comment (Shadid12) [reply langchain/document_loaders/fauna.py:49]: Yes this is pagination. Following is an example how to use it. I have also added this in the docs in my latest commit.

```python
query = """
Item.paginate("hs+DzoPOg ... aY1hOohozrV7A")
Item.all()
"""
loader = FaunaLoader(query, field, secret)
```

You get a `after` value if there are more data. You can get values after the curcor by passing in the `after` string in query. 

To learn more following [this link](https://fqlx-beta--fauna-docs.netlify.app/fqlx/beta/reference/schema_entities/set/static-paginate)

File: langchain/document_loaders/fauna.py (hunk 3 above)
Comment (Shadid12) [reply langchain/document_loaders/fauna.py:39]: This is needed because Fauna uses a encoding to map the database types. 
https://github.com/fauna/fauna-python/blob/main/fauna/encoding/encoder.py#L49

File: langchain/document_loaders/fauna.py (hunk 2 above)
Comment (dev2049) [reply langchain/document_loaders/fauna.py:49]: link seems to suggest you could just do (replace Item -> Set, drop Item.all())
```python
query = """
Set.paginate("hs+DzoPOg ... aY1hOohozrV7A")
"""
```
is that correct?

Comment (Shadid12) [reply langchain/document_loaders/fauna.py:49]: Yes you are correct. Updated the docs

File: tests/integration_tests/document_loaders/test_fauna.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import unittest
```
Comment (eyurtsev) [inline tests/integration_tests/document_loaders/test_fauna.py:1]: 1) Could you add `fauna` as an optional to the main group, and register `fauna` in the extended testing group
2) Add a unit test that uses @pytest.mark.parameterize and attempts to import the fauna loader (or even better uses a fixture to test the glue code between the fauna client and the results of the loader)

File: langchain/document_loaders/fauna.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+class FaunaLoader(BaseLoader):
+    def __init__(
+        self,
+        query: str,
+        page_content_field: str,
+        secrect: str,
+        metadata_fields: Optional[List[str]] = None,
+        after: Optional[str] = None,
+    ):
```
Comment (eyurtsev) [inline langchain/document_loaders/fauna.py:15]: Consider documenting these fields to make it easier for users to use it

File: langchain/document_loaders/fauna.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+from langchain.document_loaders.base import BaseLoader
+
+
+class FaunaLoader(BaseLoader):
+    def __init__(
+        self,
+        query: str,
+        page_content_field: str,
+        secrect: str,
+        metadata_fields: Optional[List[str]] = None,
```
Comment (eyurtsev) [inline langchain/document_loaders/fauna.py:13]: Favor Sequence over List for inputs to promise immutability of collection

```suggestion
        metadata_fields: Optional[Sequence[str]] = None,
```

**Review by eyurtsev — COMMENTED** [review]
@Shadid12 Looks pretty good to me! Thank you for the contribution it's much appreciated! I left a few minor comments

File: tests/integration_tests/document_loaders/test_fauna.py (hunk 4 above)
Comment (Shadid12) [reply tests/integration_tests/document_loaders/test_fauna.py:1]: updated. please review

File: pyproject.toml [hunk 7]
```diff
@@ -120,6 +120,7 @@ pytest = "^7.3.0"
 pytest-cov = "^4.0.0"
 pytest-dotenv = "^0.5.2"
 duckdb-engine = "^0.7.0"
+fauna= "0.6.1"
```
Comment (eyurtsev) [inline pyproject.toml:123]: Only packages that help with testing infrastructure go in the test group. 

```suggestion
```

See information here on how to introduce a new optional dependency: https://github.com/hwchase17/langchain/blob/master/.github/CONTRIBUTING.md#working-with-optional-dependencies

File: tests/integration_tests/document_loaders/test_fauna.py (hunk 4 above)
Comment (eyurtsev) [reply tests/integration_tests/document_loaders/test_fauna.py:1]: Hi @Shadid12!

Thanks for the unit test!

---

Please see instructions here on how to register in the optional dependency: 

https://github.com/hwchase17/langchain/blob/master/.github/CONTRIBUTING.md#working-with-optional-dependencies

The dependency needs to be added to the `extended_testing`  extra in pyproject.toml

---

langchain relies on the pytest framework for unit testing -- classes should not inherit from `unittest.TestCase` -- there might be a few exceptions to this in the code base, but most of the code is relying on pytest style testing

**Review by eyurtsev — CHANGES_REQUESTED** [review]

File: langchain/document_loaders/fauna.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        after: Optional[str] = None,
+    ):
+        self.query = query
+        self.page_content_field = page_content_field
+        self.secrect = secrect
+        self.metadata_fields = metadata_fields
+
+    def load(self) -> List[Document]:
```
[the remark this answers is not in the document]
Comment (eyurtsev) [reply langchain/document_loaders/fauna.py:29]: For a lazy implementation, we want to implement `lazy_load` rather than `load` -- take a look at the base class to see the method signatures with docs that explain what's going on! :)
